### PR TITLE
Convert Gradle to Maven Build

### PR DESCRIPTION
**Pull request by ALRubinger**
# Summary

This PR creates a mechanism whereby we may align dependencies between `web5-kt` and projects built atop it, (ie. `tbdex-kt`). This satisfies #217. It does so by replacing the existing Gradle build with a Maven one and adding explicit `<dependencyManagement>` sections through which dependency versioning can be governed and exported for use in other projects. A more in-depth account of that reasoning is [here on Discord](https://discord.com/channels/937858703112155166/1218077520411234304).

This is offered as a drop-in replacement for the existing build:
* No code changes to the SDK itself
* Feature parity with the existing build
* Updates to unify the release process alongside @leordev pipelines applied throughout the rest of the SDKs.
* Updated `README` showing how to use all aspects of the local build, publishing, and release process

Expected impact for devs of `web5-kt` is in 3 dimensions:
* `mvn verify` is now the command to run the build locally
* The "Release and Publish" workflow kicks off the release process, taking one input for the version to release.
* Dependency management is now done via Maven with docs in the `pom.xml` files outlining how to do this.

## API Docs

* Ported.
* Added instructions to run locally to `README`.
* Tested deployment; available at: https://alrubinger.github.io/web5-kt/docs/
* Set to trigger on release, not push to `main`. 

## Kover and CodeCov

* Ported
* Kover creates reports; uploaded to CodeCov for inspection
* Verifiable as working here: https://app.codecov.io/github/TBD54566975/web5-kt/tree/issue-217%2Fmaven-build

## Detekt

* Ported
* Verified tied into the Maven `verify` lifecycle, for instance when running the build via `mvn verify`:

```
[INFO] --- detekt:1.23.5:check (default) @ web5-common ---
[INFO] Args:
.........

9 kotlin files were analyzed.
Complexity Report:
	- 941 lines of code (loc)
	- 590 source lines of code (sloc)
	- 425 logical lines of code (lloc)
	- 256 comment lines of code (cloc)
	- 122 cyclomatic complexity (mcc)
	- 71 cognitive complexity
	- 0 number of total code smells
	- 43% comment source ratio
	- 287 mcc per 1,000 lloc
	- 0 code smells per 1,000 lloc

Project Statistics:
	- number of properties: 97
	- number of functions: 36
	- number of classes: 12
	- number of packages: 1
	- number of kt files: 9
```

## Release

* Now done through one workflow: "Release and Publish"
* Accepts a single input for the version to release
* Keeps the Maven project at `0.0.0-main-SNAPSHOT`, effectively unversioning upstream dev as we've done in Gradle.
* Triggering the "Release and Publish" workflow builds, tests, publishes to TBD Artifactory, tags in `git` as `v$version`, publishes to Maven Central, creates a GitHub Release, and publishes API Docs.
* @jiyoontbd would like a selection screen like we do in web5-swift; picking whether the next release should be a patch, minor, or major bump over the latest one. I've got all the logic given a release name here in this PR; will leave to @leordev afterwards to make this UX enhancement to assemble the next version based on prior tags.
* Input screen:
* image
* This will kick off the full publishing pipeline:
* image

### Release on TBD Artifactory
This is triggered by the "Release and Publish" workflow. This example has version: `0.15.0-issue217-alpha-1`

* https://blockxyz.jfrog.io/artifactory/tbd-oss-releases-maven2/xyz/block/

### GitHub Release

Same as the example above: a GitHub Release for: `0.15.0-issue217-alpha-1`

* https://github.com/ALRubinger/web5-kt/releases/tag/v0.15.0-issue217-alpha-1

### SNAPSHOT on TBD Artifactory
These are published on every push to main. This example has version: `commit-2ed97da-SNAPSHOT`

* https://blockxyz.jfrog.io/artifactory/tbd-oss-snapshots-maven2/xyz/block/

### Release Candidate Build on TBD Artifactory

Published: https://blockxyz.jfrog.io/artifactory/tbd-oss-releases-maven2/xyz/block/web5/1.0.0-maven-rc-1/

### Release on Maven Central

* To do this we'd need to cut an actual release. :) Let's trust this works and if it doesn't we'll fix up the workflows after merge.

### CODE REVIEW

**Comment by ALRubinger** [comment]
Update on how this is going @jiyoontbd @frankhinek:

* The build and tests are passing for the Web5 suite.
* Security Vulns are clear
* The POM we get out of this will be able to be imported into `tbdex-kt` so the two projects share versioning and make a cohesive platform.

Still TODO:
* Hook in Dokka
* Get the vector tests running
* Set up Maven publishing
* ...and anything else custom in the existing Gradle configs need to be ported over.

But this is an encouraging stopping point for today. Thanks for resolving that cyclic dependency. Found no other issues in Mavenizing. Looks like I won't need much from @jiyoontbd until we do review of this chonky PR. 🤘🏻

**Comment by ALRubinger** [comment]
Things that will require some validation and perhaps follow-on commits after this is merged:

* Publish Action
* Publish SNAPSHOT Action

These we can't fully test until they're in `main`, but in case of errors we can fix them up after. I've done my best to validate the commands locally first.

**Comment by ALRubinger** [comment]
Added Dokka config using Dokka CLI as Dokka Maven plugin does not support multimodule builds. The config looks a little different from before, with no sidebar. Gotta be "good enough" for now, though I'm very open to anyone who can restore the left nav sidebar.
image

**Comment by leordev** [comment]
> Added Dokka config using Dokka CLI as Dokka Maven plugin does not support multimodule builds. The config looks a little different from before, with no sidebar. Gotta be "good enough" for now, though I'm very open to anyone who can restore the left nav sidebar. 

Does the sidebar shows up when one of the modules is selected? If so, I think we should be ok for now, since they are different packages. Are there any interlinked docs? For example, in the dids package, a link to a crypto class/definition that does not work anymore?

**Comment by ALRubinger** [comment]
> > Added Dokka config using Dokka CLI as Dokka Maven plugin does not support multimodule builds. The config looks a little different from before, with no sidebar. Gotta be "good enough" for now, though I'm very open to anyone who can restore the left nav sidebar.
> 
> Does the sidebar shows up when one of the modules is selected? If so, I think we should be ok for now, since they are different packages. Are there any interlinked docs? For example, in the dids package, a link to a crypto class/definition that does not work anymore?

@leordev It doesn't, boo. I revised the `README` to include "Generating API Docs Locally" if you wanna give it a spin. Note the Hermit installation and `source ./bin/activate-hermit` steps which are new in `main` and I've rebased into here.

**Comment by ALRubinger** [comment]
Update: Getting close. Need to reconcile how we publish to Maven Central now. All publishing in this PR is to TBD Artifactory as it stands. Will link with @leordev on that tomorrow.

**Comment by ALRubinger** [comment]
I think we have to run Kover to generate the reports that CodeCov reads, good catch from @leordev

**Comment by ALRubinger** [comment]
Doesn't look like the cache on the `build-test-deploy` script is doing anything; lots of Maven Central downloads there. Why?

**Comment by ALRubinger** [comment]
> Things that will require some validation and perhaps follow-on commits after this is merged:
> 
> * Publish Action
> * Publish SNAPSHOT Action
> 
> These we can't fully test until they're in `main`, but in case of errors we can fix them up after. I've done my best to validate the commands locally first.

This is no longer true. I've updated the workflows here to run on this branch and they're tested on this PR thanks to @leordev showing me how to do that.

**Comment by ALRubinger** [comment]
Boom Kover works now.

https://app.codecov.io/github/TBD54566975/web5-kt/tree/issue-217%2Fmaven-build

image

**Comment by ALRubinger** [comment]
For testing of GitHub Actions, I'm continuing work on this branch in my fork on `main`; will bring back here after I can validate Actions are working correctly.

https://github.com/ALRubinger/web5-kt/

File: .github/workflows/build-test-publish-tbd-artifactory.yml
Comment (leordev) [inline .github/workflows/build-test-publish-tbd-artifactory.yml:145]: nit: cleanup

File: .github/workflows/build-test-publish-tbd-artifactory.yml
Comment (leordev) [inline .github/workflows/build-test-publish-tbd-artifactory.yml:25]: You removed the trigger on every PR branch. This way our only CI point won't run for PRs. Here we should add back the pull request trigger and skip the snapshot publication to artifactory... (or publish to artifactory anyway since its a snapshot)

Comment (ALRubinger) [reply .github/workflows/build-test-publish-tbd-artifactory.yml:25]: Oh great catch @leordev would you mind adding those back? Sure, publish to artifactory SNAPSHOTs anyway because then folks could consume PRs to test, could be helpful.

File: .github/workflows/build-test-publish-tbd-artifactory.yml
Comment (ALRubinger) [reply .github/workflows/build-test-publish-tbd-artifactory.yml:145]: Thank you! I will add a commit to address.

File: .github/workflows/build-test-publish-tbd-artifactory.yml
Comment (leordev) [reply .github/workflows/build-test-publish-tbd-artifactory.yml:25]: Done!

File: .github/workflows/build-test-publish-tbd-artifactory.yml
Comment (leordev) [reply .github/workflows/build-test-publish-tbd-artifactory.yml]: Dont worry! I addressed it as part of the below fix!

File: pom.xml
Comment (leordev) [inline pom.xml:10]: Question: By having the `web5-parent` here, are we publishing any `web5-parent` package? Or it's just used for gluing the subpackages as I've seen references to parent across the board?

**Review by leordev — APPROVED** [review]

File: pom.xml
Comment (ALRubinger) [reply pom.xml:10]: Yep, we publish the `web5-parent`; it'll be used as the parent of `tbdex-parent`, thereby slimming down all the plugin configs, versioning info, etc necessary at the tbDEX level. Like [this](https://github.com/TBD54566975/tbdex-kt/blob/issue-204/maven-build/pom.xml#L13).

File: .github/workflows/ci.yml
Comment (jiyoontbd) [inline .github/workflows/ci.yml:44]: can these have names too?

File: .github/workflows/ci.yml
Comment (jiyoontbd) [inline .github/workflows/ci.yml:74]:
```suggestion
      - name: Init Hermit
      - uses: cashapp/activate-hermit@v1
        with:
          cache: true
```

i meant to do this in kt yml files, but we can use hermit cache like this https://github.com/TBD54566975/web5-go/blob/16aa6fd03c18edf83f359cf6db6b4235e33cab0b/.github/actions/setup/action.yml#L6

my suggestion indent is probably wrong

File: .github/workflows/ci.yml
Comment (jiyoontbd) [inline .github/workflows/ci.yml:125]: chad tells me sign-artifacts is a custom profile. can you point to where that is and what it does?

File: .github/workflows/ci.yml
Comment (jiyoontbd) [inline .github/workflows/ci.yml:64]: why did we get rid of `&& matrix.os == 'ubuntu-latest'` ?

File: .github/workflows/ci.yml
Comment (jiyoontbd) [inline .github/workflows/ci.yml:73]: same here

File: .github/workflows/release.yml
Comment (jiyoontbd) [inline .github/workflows/release.yml:34]: you can use hermit cache here too

File: README.md
Comment (jiyoontbd) [inline README.md:236]: does this mean we'll create git tags like `v0.11.0-somesha-SNAPSHOT` ?

Comment (jiyoontbd) [reply README.md:236]: oh wait maybe not, this is under the release section

File: common/pom.xml
Comment (jiyoontbd) [inline common/pom.xml:9]: can you walk me through how this version value would change? who changes it, me, or the ci?

File: common/pom.xml
Comment (jiyoontbd) [inline common/pom.xml:25]: did you mean to have this comment inside `<properties>` section?

File: config/dokka-configuration.json
Comment (jiyoontbd) [inline config/dokka-configuration.json:1]: feels a bit weird to have this live inside config folder, how do you feel about moving it to live next to dokka.sh in `scripts`? or maybe `scripts/config` ?

File: credentials/pom.xml
Comment (jiyoontbd) [inline credentials/pom.xml:41]: 1. where is the project.version declared? (i'm gonna guess at top level pom.xml)
2. unlike JS, we'll bump the version across all packages, even if we update one package, right?

File: jitpack.yml
Comment (jiyoontbd) [inline jitpack.yml:1]: what was this used for 🤔

File: pom.xml
Comment (jiyoontbd) [inline pom.xml:59]: what does this block do? i'm hoping this doesn't mean that testing is bundled in the web5 package.

File: pom.xml
Comment (jiyoontbd) [inline pom.xml:111]: `assertj` is a test dependency, no?

File: pom.xml
Comment (jiyoontbd) [inline pom.xml:347]: think we can get rid of these comments, since over time they'll become stale

File: pom.xml
Comment (jiyoontbd) [inline pom.xml:369]:
```suggestion
  <!-- Test Dependencies -->
```
?

File: pom.xml
Comment (jiyoontbd) [inline pom.xml:608]: git commands but in xml 😂

Comment (jiyoontbd) [reply pom.xml:608]: how was this done in gradle?

File: .gitignore
Comment (jiyoontbd) [inline .gitignore:9]: we can get rid of gradle gitignores now, yea?

File: .github/workflows/ci.yml
Comment (ALRubinger) [reply .github/workflows/ci.yml:28]: Yep, this is for a pull request targeted to be merged to `main`, so yep it'll run on PRs :)

File: .github/workflows/ci.yml
Comment (ALRubinger) [reply .github/workflows/ci.yml:44]: @leordev says he'll polish to add names; thanks!

File: .github/workflows/ci.yml
Comment (ALRubinger) [reply .github/workflows/ci.yml:125]: Yep it's in the root/parent `pom.xml` and it does GPG signing of the artifacts using the TBD key stored in secrets.

File: .github/workflows/ci.yml
Comment (ALRubinger) [reply .github/workflows/ci.yml:64]: Because this uploading only runs in the context of the ubuntu job, so we don't need to check anymore. It used to be a matrix build, so we had to check. Now, we don't upload in the macOS build at all, so no need to check.

**Comment by ALRubinger** [comment]
Add to `README` - Snapshots are published using version format: `commit-{shortSHA}-SNAPSHOT`. @leordev will update.

File: .github/workflows/ci.yml
Comment (leordev) [reply .github/workflows/ci.yml:73]: same as above!

File: credentials/pom.xml
Comment (leordev) [reply credentials/pom.xml:41]: 1. Yes.
2. Correct, at least this is how we are doing with gradle so far too.

File: jitpack.yml
Comment (leordev) [reply jitpack.yml:1]: I think it was being used when we used to publish to jitpack registry

File: README.md
Comment (leordev) [reply README.md:236]: Correct, git tags are only for releases

File: pom.xml
Comment (mistermoe) [inline pom.xml:2]: lewl the maven URLs lead to 404s

File: common/pom.xml
Comment (leordev) [reply common/pom.xml:9]: The release CI will bump it with the desired release version. We won't bump or keep the current release here

File: .gitignore
Comment (leordev) [reply .gitignore:9]: We could, but since its not harmful I'd rather leave it here... Also especially since existing devs when pull this new version might have their gradle folders and commit them by accident (not biggie, just a nit)

File: pom.xml
Comment (leordev) [reply pom.xml:608]: Seems that gradle didnt need it, its a `maven-release-plugin` thing according to the comments above

File: .github/workflows/ci.yml
Comment (ALRubinger) [reply .github/workflows/ci.yml:125]: The workflow sets the env vars as expected by the Maven Plugin doing the signing: https://www.simplify4u.org/sign-maven-plugin/sign-mojo.html

File: pom.xml
Comment (ALRubinger) [reply pom.xml:59]: This tells the parent POM which builds to run. 

The rollup you're looking for is the `distribution` module, which then in turn defines the deps in the rollup:
https://github.com/TBD54566975/web5-kt/blob/issue-217/maven-build/distribution/pom.xml#L33

Comment (ALRubinger) [reply pom.xml:59]: `<module>` tells Maven which submodules are part of the main build.

File: pom.xml
Comment (ALRubinger) [reply pom.xml:111]: OK, great catch let's move `org.assertj` to the test deps section

File: pom.xml
Comment (ALRubinger) [reply pom.xml:347]: @jiyoontbd syas let's leave here because, yes, still relevant as of now - and future maintainers will know if it's safe to remove and why

File: pom.xml
Comment (ALRubinger) [reply pom.xml:369]: Reviewed w/ @jiyoontbd - these are the common deps for all submodules, both test and runtime

File: pom.xml
Comment (ALRubinger) [reply pom.xml:608]: Maven plugin for Release does not handle submodules so we need to hack as documented above this section to get the submodules checked out before a release build.

File: .gitignore
Comment (ALRubinger) [reply .gitignore:9]: @jiyoontbd and I agree w/ Leo, let's leave

File: pom.xml
Comment (ALRubinger) [reply pom.xml:2]: Blame Maven @mistermoe hahahaha: https://maven.apache.org/pom.html

File: .github/workflows/ci.yml
Comment (leordev) [reply .github/workflows/ci.yml:74]: I've added and confirmed in their source that they use the gh cache under the hood, check it out:
https://github.com/cashapp/activate-hermit/blob/v1/action.yml#L21-L27

cc @ALRubinger to replicate this on tbdex-kt ci!

Comment (ALRubinger) [reply .github/workflows/ci.yml:69]: Noted; will replicate! TY for handling here @leordev and for the improvement suggestion @jiyoontbd!

File: .github/workflows/ci.yml
Comment (jiyoontbd) [inline .github/workflows/ci.yml:144]: you can remove this comment now :)

Comment (leordev) [reply .github/workflows/ci.yml]: Removed! cc @ALRubinger

**Review by jiyoontbd — APPROVED** [review]
🥳 🌟 🎉 🥇

**Pull request merged**
